js/src/instructions/withdraw.ts

```diff
@@ -0,0 +1,102 @@
+import * as BufferLayout from '@solana/buffer-layout';
+import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
+import {
+  PublicKey,
+  StakeProgram,
+  SystemProgram,
+  SYSVAR_CLOCK_PUBKEY,
+  TransactionInstruction,
+} from '@solana/web3.js';
+import BN from 'bn.js';
+import type { Lamports, ProgramAddresses, Snapshot } from '../types';
+import {
+  findAuthorityProgramAddress,
+  getHeaviestValidatorStakeAccount,
+} from '../utils';
+
+export const getWithdrawInstruction = async (
+  snapshot: Snapshot,
+  programAddresses: ProgramAddresses,
```

Comment:
Let’s make `ProgramAddresses` a field of `Snapshot` and initialize it when we create the snapshot. That’s one less thing for callers to worry about, and also makes it impossible to mix a snapshot created from say testnet `ProgramAddresses` with mainnet `ProgramAddresses` here.